File: java/2452.java
class Solution {
    class Trie {
        Trie c[] = new Trie[26];
    }

    Trie r;

    private void makeTrie(String s) {
        Trie t = r;
        for (int c : s.toCharArray()) {
            if (t.c[c - 97] == null)
                t.c[c - 97] = new Trie();
            t = t.c[c - 97];
        }
    }

    private boolean canMake(Trie t, char w[], int i, int n, int l) {
        if (t.c[w[i] - 97] != null && (i == n - 1 || canMake(t.c[w[i] - 97], w, i + 1, n, l)))
            return true;
        if (l == 0)
            return false;
        for (Trie h : t.c)
            if (h != null && (i == n - 1 || canMake(h, w, i + 1, n, l - 1)))
                return true;
        return false;
    }

    public List<String> twoEditWords(String[] q, String[] d) {
        r = new Trie();
        for (String s : d)
            makeTrie(s);
        List<String> l = new ArrayList<>();
        for (String s : q)
            if (canMake(r, s.toCharArray(), 0, s.length(), 2))
                l.add(s);
        return l;
    }
}

File: c/2452.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
struct Trie {
    struct Trie* c[26];
};

typedef struct Trie Trie;

Trie* r = NULL;

Trie* new_Trie() {
    Trie* t = (Trie*)malloc(sizeof(Trie));
    for (int i = 0; i < 26; i++)
        t->c[i] = NULL;
    return t;
}

void makeTrie(char* s) {
    Trie* t = r;
    int n = strlen(s);
    for (int i = 0; i < n; i++) {
        if (t->c[s[i] - 97] == NULL)
            t->c[s[i] - 97] = new_Trie();
        t = t->c[s[i] - 97];
    }
}

bool canMake(Trie* t, char* w, int i, int n, int l) {
    if (t->c[w[i] - 97] != NULL &&
        (i == n - 1 || canMake(t->c[w[i] - 97], w, i + 1, n, l)))
        return true;
    if (l == 0)
        return false;
    for (int j = 0; j < 26; j++)
        if (t->c[j] != NULL &&
            (i == n - 1 || canMake(t->c[j], w, i + 1, n, l - 1)))
            return true;
    return false;
}

char** twoEditWords(char** q, int qs, char** d, int ds, int* rs) {
    r = new_Trie();
    for (int i = 0; i < ds; i++)
        makeTrie(d[i]);
    int c = 0;
    for (int i = 0; i < qs; i++)
        if (canMake(r, q[i], 0, strlen(q[i]), 2))
            c++;
    *rs = c;
    char** res = (char**)malloc(sizeof(char*) * c);
    int j = 0;
    for (int i = 0; i < qs; i++)
        if (canMake(r, q[i], 0, strlen(q[i]), 2))
            res[j++] = q[i];
    return res;
}
